### ISSUE TITLE
Invalid URL after lang switch

### ISSUE THREAD

**Issue by btravel** [keycloak/keycloak#16063]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

authentication

### Describe the bug

I'm using oidc authentication with a standard flow (public client with a keycloak-js adapter in SPA).

After my web app redirects to a keycloak login form I got URL like:

/realms/.../protocol/openid-connect/auth?client_id=...&redirect_uri=...&state=...&response_mode=fragment&response_type=code&scope=openid&nonce=...&code_challenge=...&code_challenge_method=S256

I'm using a custom authenticator with a custom login form rendered by context.challenge(...).createForm(...);  If I try to switch a language in this form my URL replaced with /realms/.../protocol/openid-connect/auth?client_id=...&tab_id=...&execution=...&kc_locale=en

So, I lost redirect_uri and got an error "Invalid parameter: redirect_uri"

### Version

20.0.2

### Expected behavior

Language switch buttons only change a locale in URL

### Actual behavior

Language switch buttons triggers URL replace,  redirect_uri and other parameters are lost

### How to Reproduce?

Using custom authenticator switch language in a custom login form rendered by context.challenge(...).createForm(...)

**Comment by mposolda** [comment]
I am adding to Backlog as I am not sure if Keycloak team will have opportunity to look at this in the near future. Everyone is welcome to investigate and send PR for this. Please take a look at https://www.keycloak.org/community to the "Contributors" section for the details.

**Comment by davidklika** [comment]
I face the same issue. I find out, that 
the default browser login:
 * uses login-actions/authenticate endpoint 
 * language switch links contains only a few query parameters (client_id, tab_id, kc_locale)
 * the language switch works as expected there

my custom flow:
 * is accessed using the openid-connect/auth endpoint
 * query parameters are the same
 * the language switch stops very early at AuthorizationEndpoint.process() line checker.checkRedirectUri();

So the query parameters set is not valid for the openid-connect/auth endpoint. (Not sure if the same issue holds for nay of DeleteAccount / TermsAndConditions / UpdateEmail / WebAuthnRegister that also use LoginFormsProvider.createForm).

Query parameters set is built in LocaleBean using an UriBuilder passed from FreeMarkerLoginFormsProvider.createCommonAttributes or FreeMarkerLoginFormsProvider.createForm. I think that either it could include all query parameters from the request or a createForm method variant could provide a way how to specify additional parameters.
Any opinions?

**Comment by Ketec** [comment]
This is a problem, 
it seems calling context.success() does not update the URL like a manual action through submit button would. Reloading the page also fixes it.

In my case - i'm skipping username view if it is already provided via login_hint.

submit calls the actionurl  - but calling success() keeps the existing uri and does not update urlbean

**Comment by Sandsten** [comment]
I'm affected by this issue as well. 

It only occurs if the custom template is rendered by the first authentication step a user lands on.

**Comment by danielFesenmeyer** [comment]
We have also observed this behavior.

As described by @Sandsten, when a custom page is the first page in the authentication flow. For this case, the suggested workaround in the discourse group worked quite well: https://keycloak.discourse.group/t/switching-locale-results-in-error-missing-parameter-response-type/7182. But it doesn't work with more than one custom form in a row and is cumbersome to maintain, when several custom forms are implemented.

We also found that the "expired page", which is erroneously shown in such pages, does not evaluate the last selected locale of the user.

The linked PR addresses these issues: https://github.com/keycloak/keycloak/pull/41231
(Still in draft, because it's currently reviewed by a team member.)
